numpyro/infer/hmc_gibbs.py

```diff
@@ -476,48 +504,297 @@ class HMCECS(HMCGibbs):
+        model_kwargs["_gibbs_sites"] = z_gibbs
+        model_kwargs["_gibbs_state"] = gibbs_state
+        hmc_state = self.inner_kernel.sample(hmc_state, model_args, model_kwargs)
 
         z = {**z_gibbs, **hmc_state.z}
+        return HMCECSState(z, hmc_state, rng_key, gibbs_state, accept_prob)
 
-        return HMCGibbsState(z, hmc_state, rng_key)
+    @property
+    def grad_(self):
+        if not hasattr(self, '_grad_mode'):
+            if self.inner_kernel._forward_mode_differentiation:
+                grad_ = jacfwd
+            else:
+                grad_ = grad
+            self._grad_ = grad_
+        return self._grad_
+
+
+def perturbed_method(subsample_plate_sizes, proxy_fn):
+    def estimator(likelihoods, params, gibbs_state):
+        subsample_log_liks = defaultdict(float)
+        for (fn, value, name, subsample_dim) in likelihoods.values():
+            subsample_log_liks[name] += _sum_all_except_at_dim(fn.log_prob(value), subsample_dim)
+
+        log_lik_sum = 0.
+
+        proxy_value_all, proxy_value_subsample = proxy_fn(params, subsample_log_liks.keys(), gibbs_state)
+
+        for name, subsample_log_lik in subsample_log_liks.items():  # loop over all subsample sites
+            n, m = subsample_plate_sizes[name]
+
+            diff = subsample_log_lik - proxy_value_subsample[name]
+
+            unbiased_log_lik = proxy_value_all[name] + n * jnp.mean(diff)
+            variance = n ** 2 / m * jnp.var(diff)
+            log_lik_sum += unbiased_log_lik - 0.5 * variance
+        return log_lik_sum
+
+    return estimator
+
+
+def taylor_proxy(reference_params):
+    """  Control variate for unbiased log likelihood estimation using a Taylor expansion around a reference
+    parameter. Suggest for subsampling in [1].
+
+    :param reference_params: Model parameterization at MLE or MAP-estimate.
+
+    ** References: **
+
+    [1] Towards scaling up Markov chainMonte Carlo: an adaptive subsampling approach
+        Bardenet., R., Doucet, A., Holmes, C. (2014)
+    """
+
+    def construct_proxy_fn(prototype_trace, subsample_plate_sizes, model, model_args, model_kwargs, num_blocks=1):
+        ref_params = {name: biject_to(prototype_trace[name]["fn"].support).inv(value)
+                      for name, value in reference_params.items()}
+
+        ref_params_flat, unravel_fn = ravel_pytree(ref_params)
+
+        def log_likelihood(params_flat, subsample_indices=None):
+            if subsample_indices is None:
+                subsample_indices = {k: jnp.arange(v[0]) for k, v in subsample_plate_sizes.items()}
+            params = unravel_fn(params_flat)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+
+                with block(), trace() as tr, substitute(data=subsample_indices), substitute(data=params), \
+                        substitute(substitute_fn=partial(_unconstrain_reparam, params)):
```

Comment:
can we use
```
params = {name: biject_to(prototype_trace[name]["fn"].support)(value) for name, value in params.items()}
```
here so you can keep `substitute(data=params)` and remove this `substitute_fn`. I don't know how it works when we use both `data=params` and `substitute_fn=...` here.

Reply:
I'll update this.